esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -889,49 +889,96 @@ void EsdlServiceImpl::handleFinalRequest(IEspContext &context,
+                    qname = stag.getQName();
+                    if (strcmp(qname, "soap:Envelope") == 0)
+                        tmpOut << "<Dataset name=\"DesdlSoapRequestEcho\">";
+                    else if (strcmp(qname, "soap:Body") == 0)
+                        tmpOut << "<Row>";
+                    else if (queryname && strcmp(qname, queryname) == 0)
+                        tmpOut << "<roxierequest name=\"" << queryname << "\">";
+                    else if (strcmp(qname, requestname) == 0)
+                        tmpOut << "<esprequest name=\"" << requestname << "\">";
+                    else
+                    {
+                        tmpOut << '<';
+                        if (strncmp(qname, "xsdl:", 5) == 0)
+                            tmpOut << "xsdl_" << qname + 5; // eliminate problematic namespace prefix
+                        else
+                            tmpOut << qname;
+                        for (int idx = 0; idx < stag.getLength(); idx++)
+                        {
+                            encodeXML(stag.getValue(idx), encodedContent.clear());
+                            tmpOut << ' ' << stag.getRawName(idx) << "=\"" << encodedContent << '"';
+                        }
+                        tmpOut << '>';
+                    }
+                    break;
 
-            if (bodyTree)
-            {
-                auto dsTree = responseResult->addPropTree("Dataset");
-                auto rowTree = dsTree->addPropTree("Row", LINK(bodyTree));
+                case XmlPullParser::END_TAG:
+                    xpp.readEndTag(etag);
+                    qname = etag.getQName();
+                    if (strcmp(qname, "soap:Envelope") == 0)
+                        tmpOut << "</Dataset>";
+                    else if (strcmp(qname, "soap:Body") == 0)
+                        tmpOut << "</Row>";
+                    else if (queryname && strcmp(qname, queryname) == 0)
+                        tmpOut << "</roxierequest>";
+                    else if (strcmp(qname, requestname) == 0)
+                        tmpOut << "</esprequest>";
+                    else if (strncmp(qname, "xsdl:", 5) == 0)
+                        tmpOut << "</xsdl_" << qname + 5 << '>'; // eliminate problematic namespace prefix
+                    else
+                        tmpOut << "</" << qname << '>';
+                    break;
 
-                dsTree->addProp("@name", "DesdlSoapRequestEcho");
-                toXML(responseTree, out.clear());
+                case XmlPullParser::CONTENT:
+                    content = xpp.readContent();
+                    if (!isEmptyString(content))
+                    {
+                        encodeXML(content, encodedContent.clear());
+                        tmpOut << encodedContent;
+                    }
+                    break;
+                }
             }
+            tmpOut.append(echoInsertionPoint);
+            out.swapWith(tmpOut);
```

Comment:
Do you think this could have been done without tmpOut at all?  I.e. append the info directly after echoInsertionPoint and then add the trailing close tags?

Reply:
I don't want an unexpected parser exception to leave `out` in an invalid state, because I'm not convinced an error parsing this should be allowed to break the response. If there are no other exceptions possible from the parsing, then yes, `out` could be manipulated in place. I would probably copy the trailing close tags into a second string rather than trying to recreate them, because I think that would be cleaner than finding what comes between `</Results>` and `</soap:Body>`.

Reply:
If I keep `tmpOut`, but only use it for the parsed dataset to be inserted, I could then manipulate `out` by inserting `tmpOut` at the insertion point only when `tmpOut` is assembled successfully.

Reply:
ok